[NEMO-216,251,259] Support side inputs and windowing

johnyangk opened the pull request:
JIRA: [NEMO-216: Support side inputs and windowing](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-216)
(+ NEMO-251, NEMO-259)

**Major changes:**
- MaterializedViewReader (implements ReadyCheckingSideInputReader) and appropriate push-back logics
- SideInputTransform: Wraps materialized views with an index in SideInputElement to distinguish multiple sideinputs destined to the same DoFn (replaces the previous usage of BroadcastVariable ids)
- In the runtime, treat a SideInput as just another DataFetcher (do not use BroadcastManagerWorker): all sideinput-specific logics stay in the Beam frontend

**Minor changes to note:**
- Clean up watermark propagation flow for DataFetchers
- In MultiThreadParentTaskDataFetcher, let the task thread handle watermarks just like other elements and not data fetcher threads

**Tests for the changes:**
- WindowedBroadcastITCase
- DoFnTransformTest#testSideInputs

**Other comments:**
- This is the last sub-issue of the umbrella issue: [NEMO-229] Support watermark, so once this PR is merged we should be ready to mark that as resolved

taegeonum commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/SideInputTransform.java:
```diff
@@ -0,0 +1,71 @@
+ * Side input transform implementation.
+ * @param <T> input/output type.
+ */
+public final class SideInputTransform<T> implements Transform<WindowedValue<T>, SideInputElement> {
```
I think this class can be removed if we use `view` instead of `index`. We can emit the materialized data in `CreateViewTransform` with `view`.

taegeonum commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/data/partitioner/DedicatedKeyPerElementPartitioner.java:
```diff
@@ -33,6 +33,7 @@
    * Constructor.
    */
   public DedicatedKeyPerElementPartitioner() {
+    // TODO should not be 0
```
?

taegeonum commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/task/SourceVertexDataFetcher.java:
```diff
@@ -104,7 +107,8 @@ private boolean isWatermarkTriggerTime() {
     if (!bounded && isWatermarkTriggerTime()) {
-      return new Watermark(readable.readWatermark());
+      // index=0 as there is only 1 input stream
+      return new WatermarkWithIndex(new Watermark(readable.readWatermark()), 0);
```
Why do we emit this here? The index is outgoing edge index between inter task

johnyangk replied:
I'll take a look at this.

johnyangk commented:
@taegeonum I've fixed the tests and made the following changes. Please take a look. 😄 
- Filed https://issues.apache.org/jira/browse/NEMO-297 to remove SideInputTransform for more concise code. (+ performance testing)
- Changed DoFnTransform to use PushbackSideInputDoFnRunner only when there is a side input
- Changed source data fetchers to use `Watermark`s.

johnyangk commented:
@taegeonum (I think) I've also fixed the empty side input issue. `PerPercentileAverageITCase` appears to emit empty side inputs and the test passes after this commit:
https://github.com/apache/incubator-nemo/pull/159/commits/0f5c630080b66c5c5a11f70097b082acbf86b375

johnyangk commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/AbstractDoFnTransform.java:
```diff
@@ -132,10 +145,14 @@ public final DoFn getDoFn() {
     if (bundleFinished) {
       bundleFinished = false;
-      doFnRunner.startBundle();
+      if (pushBackRunner == null) {
```
I think it's better to maintain a single method for doFnRunner/pushBackRunner, as bundle-related methods are called from multiple different places.

taegeonum replied:
It looks like we use different start/finish bundle strategy for simple DoFnRunner and pushBackRunner. Separating them into different methods would be good.

johnyangk replied:
That will give us 3 duplicate methods.
I left a comment on refactoring. Maybe we can come back to this when implementing streaming partial combiner (i.e., combining ReduceFn)?

johnyangk commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/CreateViewTransform.java:
```diff
@@ -112,20 +113,12 @@ public void onWatermark(final Watermark inputWatermark) {
   public void close() {
     onWatermark(new Watermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()));
-
-    if (!isEmitted) {
```
I think so (PerPercentile test passes)
It will be handled more explicitly in https://github.com/apache/incubator-nemo/pull/159/files#diff-30a095310e20d5a90938099d206b1757R32

johnyangk commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/task/MultiThreadParentTaskDataFetcher.java:
```diff
@@ -99,35 +96,14 @@ Object fetchDataElement() throws IOException, NoSuchElementException {
     final List<CompletableFuture<DataUtil.IteratorWithNumBytes>> futures = readersForParentTask.read();
     numOfIterators = futures.size();
-
-    if (numOfIterators > 1) {
```
I've reverted the code 😄 
Considering that it makes the InputWatermarkManager come "in front" of the data fetcher, and hides indices from TaskExecutor.

Although there are disadvantages of a bit more complex code, and having to handle indices in each DataFetcher.

taegeonum commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/AbstractDoFnTransform.java:
```diff
@@ -132,25 +150,46 @@ public final DoFn getDoFn() {
+  /**
+   * Finish bundle without checking for conditions.
+   */
+  final void forceFinishBundle() {
```
This is only used for PushBackRunner.

johnyangk replied:
This is also used in AbstractDoFnTransform#close

taegeonum commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/PushBackDoFnTransform.java:
```diff
@@ -0,0 +1,160 @@
+    final List<WindowedValue<InputT>> pushedBackAgain = new ArrayList<>();
+    long pushedBackAgainWatermark = Long.MAX_VALUE;
+    for (WindowedValue<InputT> curPushedBack : curPushedBacks) {
+      final Iterable<WindowedValue<InputT>> pushedBack =
```
`checkAndInvokeBundle` here?

taegeonum commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/PushBackDoFnTransform.java:
```diff
@@ -0,0 +1,160 @@
+    for (WindowedValue<InputT> curPushedBack : curPushedBacks) {
+      final Iterable<WindowedValue<InputT>> pushedBack =
+        getPushBackRunner().processElementInReadyWindows(curPushedBack);
+      for (final WindowedValue<InputT> wv : pushedBack) {
```
`checkAndFinishBundle` here?

taegeonum commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/PushBackDoFnTransform.java:
```diff
@@ -0,0 +1,160 @@
+    // {@link org.apache.beam.runners.core.SimplePushbackSideInputDoFnRunner}
+    // caches for each bundle the side inputs that are not ready.
+    // We need to re-start the bundle to advertise the (possibly) newly available side input.
+    forceFinishBundle(); // forced
```
Why don't you just call `pushBackRunner.startBundle` before handlePushBacks and `pushBackRunner.finishBundle` after handlePushBacks?

johnyangk replied:
We need to force-finish the previous bundle as described in the comment.
Otherwise, DoFnTransformTest#testSideInputs fails.

taegeonum commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/PushBackDoFnTransform.java:
```diff
@@ -0,0 +1,160 @@
+    checkAndInvokeBundle();
+    // With the new side input added, we may be able to process some pushed-back elements.
+    final List<WindowedValue<InputT>> pushedBackAgain = new ArrayList<>();
+    long pushedBackAgainWatermark = Long.MAX_VALUE;
```
curPushBackWatermark = Long.MAX_VALUE

taegeonum replied:
Yes, but  I think we don't need the `pushedBackAgainWatermark` variable if we just set `curPushBackWatermark = Long.MAX_VALUE` here?

taegeonum commented on compiler/frontend/beam/src/main/java/org/apache/nemo/compiler/frontend/beam/transform/PushBackDoFnTransform.java:
```diff
@@ -0,0 +1,160 @@
+    // With the new side input added, we may be able to process some pushed-back elements.
+    final List<WindowedValue<InputT>> pushedBackAgain = new ArrayList<>();
+    long pushedBackAgainWatermark = Long.MAX_VALUE;
+    for (WindowedValue<InputT> curPushedBack : curPushedBacks) {
```
final

taegeonum commented on runtime/executor/src/main/java/org/apache/nemo/runtime/executor/task/SourceVertexDataFetcher.java:
```diff
@@ -104,6 +104,7 @@ private boolean isWatermarkTriggerTime() {
   private Object retrieveElement() throws NoSuchElementException, IOException {
     // Emit watermark
     if (!bounded && isWatermarkTriggerTime()) {
+      // index=0 as there is only 1 input stream
```
please remove this comment
